sphinx_gallery/recommender.py

```diff
@@ -0,0 +1,315 @@
+    """
+
+    def __init__(self, *, n_examples=5, tokenizer="raw"):
+        self.n_examples = n_examples
+        self.tokenizer = tokenizer
+
+    @staticmethod
+    def token_freqs(doc):
+        """Extract a dict mapping raw tokens from doc to their occurrences."""
+        token_generator = (tok.lower() for tok in re.findall(r"\w+", doc))
+        freq = defaultdict(int)
+        for tok in token_generator:
+            freq[tok] += 1
+        return freq
+
+    @staticmethod
+    def dict_freqs(doc):
+        """Extract a dict mapping list of tokens to their occurrences."""
+        freq = defaultdict(int)
+        for tok in doc:
+            freq[tok] += 1
+        return freq
+
+    @staticmethod
+    def dict_vectorizer(data_dict):
+        """Convert a dictionary of feature arrays into a sparse matrix.
+
+        Parameters
+        ----------
+        data_dict : dict
+            A dictionary of feature arrays, where each key corresponds to a feature
+            name, and each value is an array of feature values.
+
+        Returns
+        -------
+        X : sparse matrix
+            A sparse matrix in CSR format of shape (n_samples, n_features) where
+            n_samples is the number of samples in the dataset and n_features is the
+            total number of features across all samples.
+        """
+        feature_names = []
+        all_values = defaultdict(list)
+        for row in data_dict:
+            for feature_name, feature_value in row.items():
+                feature_names.append(feature_name)
+                all_values[feature_name].append(feature_value)
+
+        feature_names = list(set(feature_names))
+        feature_names.sort()
+        data = []
+        indices = []
+        indptr = [0]
+        for i, row in enumerate(data_dict):
+            for j, feature_name in enumerate(feature_names):
+                if feature_name in row:
+                    feature_value = row[feature_name]
+                    data.append(feature_value)
+                    indices.append(j)
+            indptr.append(len(indices))
+        X = sparse.csr_matrix(
+            (data, indices, indptr), shape=(len(indptr) - 1, len(feature_names))
+        )
+        return X
+
+    @staticmethod
+    def tfidf_transformer(X):
+        """Transform a term frequency matrix into a term frequency-inverse document
+        frequency (TF-IDF) matrix.
+
+        Parameters
+        ----------
+        X : {ndarray, sparse matrix} of shape (n_samples, n_features)
+            A term frequency matrix.
+
+        Returns
+        -------
+        X_tfidf : {ndarray, sparse matrix} of shape (n_samples, n_features)
+            A tf-idf matrix of the same shape as X.
+        """
+        FLOAT_DTYPES = (np.float64, np.float32, np.float16)
+
+        if not sparse.issparse(X):
+            X = sparse.csr_matrix(X, dtype=np.float64)
+        dtype = X.dtype if X.dtype in FLOAT_DTYPES else np.float64
```

Comment:
We might be able to simplify here because we can exactly know what input type we will have. Here, this is inspired from scikit-learn where we want to be the most general depending on the possible input a user can give us.